src/transforms/kubernetes/mod.rs

```diff
@@ -1 +1,552 @@
+                };
+            };
+
+            warn!(
+                message = "Temporary stoped watching Pod metadata.",
+                reason = %runtime_error
+            );
+
+            error = Some(runtime_error);
+
+            // Wait for bit before trying to watch again.
+            let _ = Delay::new(Instant::now() + RETRY_TIMEOUT)
+                .compat()
+                .await
+                .expect("Timer not set.");
+        }
+    }
+
+    // In the case of Deleted, we don't delete it's data, as there could still exist unprocessed logs from that pod.
+    // Not deleting it will cause "memory leakage" in a sense that the data won't be used ever
+    // again after some point, but the catch is that we don't know when that point is.
+    // Also considering that, on average, an entry occupies ~232B, so to 'leak' 1MB of memory, ~4500 pods would need to be
+    // created and destroyed on the same node, which is highly unlikely.
+    //
+    // An alternative would be to delay deletions of entrys by 1min. Which is a safe guess.
+    //
+    /// Extracts metadata from pod and updates metadata map.
+    fn update(&mut self, (pod, event): (Pod, PodEvent)) -> Option<Version> {
+        if let Some(pod_uid) = pod.metadata.as_ref().and_then(|md| md.uid.as_ref()) {
+            let uid: Bytes = pod_uid.as_str().into();
+
+            self.metadata.clear(uid.clone());
+
+            // Insert field values for this pod.
+            for (field, value) in self.fields.iter().flat_map(|fun| fun(&pod)) {
+                self.metadata
+                    .insert(uid.clone(), Box::new((field, FieldValue(value))));
+            }
+
+            trace!(message = "Pod updated.", %pod_uid);
+
+            if PodEvent::Deleted == event {
+                self.delete_queue
+                    .push_back((uid, Instant::now() + DELETE_DELAY));
+            }
+        }
+
+        Version::from_pod(&pod)
+    }
+
+    /// Checks if there are entries to be deleted.
+    fn delete_update(&mut self) {
+        let now = Instant::now();
+        while let Some((pod_uid, deadline)) = self.delete_queue.get(0) {
+            if *deadline <= now {
+                self.metadata.empty(pod_uid.clone());
+
+                trace!(message = "Pod metadata deleted.", pod_uid=?std::str::from_utf8(pod_uid.as_ref()));
+
+                self.delete_queue.pop_front();
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+struct FieldValue(Value);
+
+// Since we aren't using Eq feature in the evmap, we can impl Eq.
+impl Eq for FieldValue {}
+
+pub struct KubernetesPodMetadata {
+    metadata: ReadHandle<Bytes, Box<(Atom, FieldValue)>>,
+}
+
+impl Transform for KubernetesPodMetadata {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        let log = event.as_mut_log();
+
+        if let Some(Value::Bytes(pod_uid)) = log.get(&POD_UID) {
```

Comment:
should we make the pod_uid an option that defaults to `POD_UID`?

Reply:
Yes, we should.